### ISSUE TITLE
[问题/Issue] 章节X.X：效果不对

### ISSUE THREAD

**Issue by hackermmzz** [datawhalechina/happy-llm#145]
### 1. 遇到问题的章节 / Affected Chapter

5

### 2. 具体问题描述 / Problem Description

我按照文章给的代码使用A800进行训练，得到的效果和官网给的大相径庭，可以求助一下吗

### 3. 问题重现材料 / Reproduction Materials

import transformers
from transformers import PretrainedConfig,PreTrainedModel
import torch
import torch.nn as nn
import torch.nn.functional as F
from transformers.modeling_outputs import CausalLMOutputWithPast
import numpy as np
import math
###############################################
args={
    "dim":1024,
    "n_layers":18,
    "n_heads":16,
    "n_kv_heads":8,
    "vocab_size":8192,
    "hidden_dim":3072,
    "multiple_of":64,
    "norm_eps":1e-6,
    "max_seq_len":512,
    "dropout":0.075,
    "flash_attn":True
}
###############################################
#config
class ModelConfig(PretrainedConfig):
    model_type = "Tiny-K"
    def __init__(
            self,
            dim: int = 768, # 模型维度
            n_layers: int = 12, # Transformer的层数
            n_heads: int = 16, # 注意力机制的头数
            n_kv_heads: int = 8, # 键值头的数量
            vocab_size: int = 6144, # 词汇表大小
            hidden_dim: int = None, # 隐藏层维度
            multiple_of: int = 64, 
            norm_eps: float = 1e-5, # 归一化层的eps
            max_seq_len: int = 512, # 最大序列长度
            dropout: float = 0.0, # dropout概率
            flash_attn: bool = True, # 是否使用Flash Attention
            **kwargs,
    ):
        self.dim = dim
        self.n_layers = n_layers
        self.n_heads = n_heads
        self.n_kv_heads = n_kv_heads
        self.vocab_size = vocab_size
        self.hidden_dim = hidden_dim
        self.multiple_of = multiple_of
        self.norm_eps = norm_eps
        self.max_seq_len = max_seq_len
        self.dropout = dropout
        self.flash_attn = flash_attn
        super().__init__(**kwargs)
#RMS
class RMSNorm(nn.Module):
    def __init__(self, dim, eps=1e-8):
        super().__init__()
        self.dim = dim
        self.eps = eps
        self.gamma = nn.Parameter(torch.ones(dim))

    def forward(self, x):
        x=x*torch.rsqrt(x.pow(2).mean(-1,keepdim=True)+self.eps)
        x=x*self.gamma
        return x
#repeat_kv
def repeat_kv(x:torch.Tensor,rep:int):
    if rep==1:
        return x
    bsz,seqlen,head_n,head_dim=x.shape
    ret=x[:,:,:,None,:].expand(bsz,seqlen,head_n,rep,head_dim).reshape(bsz,seqlen,head_n*rep,head_dim)
    return ret
def precompute_freqs_cis(dim: int, end: int, theta: float = 10000.0):
    freq=1.0/(theta**(torch.arange(0,dim,2).float()/dim))
    t=torch.arange(end).float()
    x=torch.outer(t,freq).float()
    cos=torch.cos(x)
    sin=torch.sin(x)
    return cos,sin
def apply_rotary_emb(q:torch.Tensor,k:torch.Tensor,cos:torch.Tensor,sin:torch.Tensor):
    def reshape_for_broadcast(cs:torch.Tensor,x:torch.Tensor):
        dim=x.ndim
        shape=[d if i==1 or i==dim-1 else 1 for i,d in enumerate(x.shape)]
        return cs.view(shape)
    xq_r,xq_i=q.float().reshape(q.shape[:-1]+(-1,2)).unbind(-1)
    xk_r,xk_i=k.float().reshape(k.shape[:-1]+(-1,2)).unbind(-1)
    cos=reshape_for_broadcast(cos,xq_r)
    sin=reshape_for_broadcast(sin,xq_r)
    rq=xq_r*cos - xq_i*sin
    iq=xq_r*sin + xq_i*cos
    rk=xk_r*cos - xk_i*sin
    ik=xk_r*sin + xk_i*cos
    xq=torch.stack([rq,iq],dim=-1).flatten(3)
    xk=torch.stack([rk,ik],dim=-1).flatten(3)
    return xq.type_as(q),xk.type_as(k)

class Attention(nn.Module):
    def __init__(self,args:ModelConfig):
        super().__init__()
        self.n_heads=args.n_heads
        self.n_kv_heads=args.n_kv_heads
        self.head_dim=args.dim//args.n_heads
        self.n_rep = self.n_heads// self.n_kv_heads
        #
        self.wq=nn.Linear(args.dim,self.n_heads*self.head_dim,bias=False)
        self.wk=nn.Linear(args.dim,self.n_kv_heads*self.head_dim,bias=False)
        self.wv=nn.Linear(args.dim,self.n_kv_heads*self.head_dim,bias=False)
        self.wo=nn.Linear(self.n_heads*self.head_dim,args.dim,bias=False)
        #
        self.dropout=args.dropout
        self.dropout0=nn.Dropout(self.dropout)
        self.dropout1=nn.Dropout(self.dropout)
        #由于我有flash_attn的实现，这里先不实现普通的attention
        
        #
    def forward(self, x:torch.Tensor,cos:torch.Tensor,sin:torch.Tensor):
        bsz,seqlen,_=x.shape
        q,k,v=self.wq(x),self.wk(x),self.wv(x)
        #
        q=q.view(bsz,seqlen,self.n_heads,self.head_dim)
        k=k.view(bsz,seqlen,self.n_kv_heads,self.head_dim)  
        v=v.view(bsz,seqlen,self.n_kv_heads,self.head_dim)
        #嵌入位置编码
        q,k=apply_rotary_emb(q,k,cos,sin)
        #拓展维度
        k=repeat_kv(k,self.n_rep)
        v=repeat_kv(v,self.n_rep)
        #使用flash attention
        output=nn.functional.scaled_dot_product_attention(
            q,k,v,
            attn_mask=None,
            dropout_p=self.dropout if self.training else 0.0,
            is_causal=True
        )
        output=output.transpose(1,2).contiguous().view(bsz,seqlen,-1)
        output=self.wo(output)
        output=self.dropout1(output)
        return output
    
class MLP(nn.Module):
    def __init__(self,hidden_dim:int,dim:int,dropout:float):
        super().__init__()
        self.fc1=nn.Linear(dim,hidden_dim,bias=False)
        self.fc2=nn.Linear(hidden_dim,dim,bias=False)
        self.fc3=nn.Linear(dim,hidden_dim,bias=False)
        self.dropout=nn.Dropout(dropout)
    def forward(self,x:torch.Tensor):
        x=self.fc2(F.silu(self.fc1(x))*self.fc3(x))
        return self.dropout(x)

class DecodeLayer(nn.Module):
    def __init__(self,args:ModelConfig):
        super().__init__()
        self.attn=Attention(args)
        self.norm1=RMSNorm(args.dim,eps=args.norm_eps)
        self.norm2=RMSNorm(args.dim,eps=args.norm_eps)
        self.fnn=MLP(
            hidden_dim=args.hidden_dim,
            dim=args.dim,
            dropout=args.dropout
        )
    def forward(self,x:torch.Tensor,cos:torch.Tensor,sin:torch.Tensor):
        x1=self.norm1(x)
        x2=x+self.attn(x1,cos,sin)
        x3=x2+self.fnn(self.norm2(x2))
        return x3

class Transformer(PreTrainedModel):
    config_class = ModelConfig  # 配置类
    def __init__(self, args:ModelConfig):
        super().__init__(args)
        self.args=args
        self.vocab_size = args.vocab_size
        self.n_layers = args.n_layers
        self.token_emb=nn.Embedding(args.vocab_size,args.dim)
        #
        self.dropout=nn.Dropout(args.dropout)
        #decode
        self.layers=nn.ModuleList()
        for i in range(args.n_layers):
            self.layers.append(DecodeLayer(args))
        # final norm
        self.norm=RMSNorm(args.dim,eps=args.norm_eps)
        #输出层
        self.output=nn.Linear(args.dim,args.vocab_size,bias=False)
        # 将词嵌入层的权重与输出层的权重共享
        self.token_emb.weight = self.output.weight 
        #计算cos和sin
        cos,sin=precompute_freqs_cis(
            dim=args.dim//args.n_heads,
            end=args.max_seq_len
        )
        self.register_buffer("freqs_cos", cos, persistent=False)
        self.register_buffer("freqs_sin", sin, persistent=False)
        #初始化所有权重
        self.apply(self._init_weights)
        for pn, p in self.named_parameters():
            if pn.endswith('w3.weight') or pn.endswith('wo.weight'):
                torch.nn.init.normal_(p, mean=0.0, std=0.02/math.sqrt(2 * args.n_layers))
        #
        self.OUT = CausalLMOutputWithPast()  # 输出容器
        self._no_split_modules = [name for name, _ in self.named_modules()]  # 不分割的模块列表
    #初始化权重
    def _init_weights(self, module):
        if isinstance(module,nn.Linear):
            torch.nn.init.normal(module.weight,mean=0.0,std=0.02)
            if module.bias is not None:
                torch.nn.init.zeros_(module.bias)
        elif isinstance(module,nn.Embedding):
            torch.nn.init.normal(module.weight,mean=0.0,std=0.02)
    def forward(self,tokens=None,targets=None,input_ids=None,attention_mask=None,labels=None,**kwargs):
        # HF/PEFT 兼容：input_ids => tokens
        if tokens is None:
            tokens = kwargs.get("input_ids", None)

        # HF/PEFT 兼容：labels => targets
        if targets is None:
            targets = kwargs.get("labels", None)

        if tokens is None:
            raise ValueError("tokens (input_ids) not provided")
        #
        bsz,seqlen=tokens.shape
        x=self.token_emb(tokens)
        h=self.dropout(x)
        cos=self.freqs_cos[:seqlen,:]
        sin=self.freqs_sin[:seqlen,:]
        #通过decoder层
        for layer in self.layers:
            h=layer(h,cos,sin)
        h=self.norm(h) 
        #推理
        last_loss=None
        if targets is not None:
            logits=self.output(h)
            last_loss=F.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1), ignore_index=0, reduction='none')
        else:
            logits=self.output(h[:, [-1], :]) 
        
        self.OUT=CausalLMOutputWithPast(
            logits=logits,          # 必需：模型输出logits
            past_key_values=None,  # 必需：注意力缓存（None也可以）
            loss=last_loss               # 训练时必需：标量平均损失
        )
        return self.OUT
    @torch.inference_mode()
    def generate(self,idx,stop_id=None,max_new_tokens=256,temperature=1.0,top_k=None):
        index=idx.shape[1]
        for _ in range(max_new_tokens):
            idx_cond = idx if idx.size(1) <= self.args.max_seq_len else idx[:, -self.args.max_seq_len:]
            logits=self.forward(tokens=idx_cond).logits[:,-1,:]
            if temperature == 0.0:
                # 选择最有可能的索引
                _, idx_next = torch.topk(logits, k=1, dim=-1)
            else:
                # 缩放 logits 并应用 softmax
                logits = logits / temperature
                if top_k is not None:
                    v, _ = torch.topk(logits, min(top_k, logits.size(-1)))
                    logits[logits < v[:, [-1]]] = -float('Inf')
                probs = F.softmax(logits, dim=-1)
                idx_next = torch.multinomial(probs, num_samples=1)
            if idx_next == stop_id:
                break
            #拼接
            idx = torch.cat((idx, idx_next), dim=1)
        return idx[:,index:]
if __name__ == "__main__":
    
    config=ModelConfig(**args)
    x = torch.randint(0, 6144, (1, 50)) # [bs, seq_len]
    # 实例化LLaMA2Model
    model = Transformer(args=config)
    # 计算model的全部参数
    num_params = sum(p.numel() for p in model.parameters())
    print('Number of parameters:', num_params)

    out = model(x)
    print(out.logits.shape) # [batch_size, 1, vocab_size]

### 确认事项 / Verification

- [x] 此问题未在过往Issue中被报告过 / This issue hasn't been reported before
